In [1]:
import numpy as np 
import scipy 
import matplotlib.pyplot as plt

# РЕШЕНИЕ СИСТЕМ ЛИНЕЙНЫХ АЛГЕБРАИЧЕСКИХ УРАВНЕНИЙ ИТЕРАЦИОННЫМИ МЕТОДАМИ
# Вариант 24 
 ## $$ Ax = b $$ 
 

In [2]:
A= [ [118.8, -14 ,-5 ,-89.1],[-14.85, -20, -5 ,0 ], [ 297,16, 320, 0], [0,6, -30, -36.3 ]  ]
b= [943, -80.7, 2602.8, 1.1] 


###  Матрица моего варианта: 

In [3]:
for i in range(len(A)):
    print("%.1f "*len(A[i]) % tuple(A[i]) )
 

118.8 -14.0 -5.0 -89.1 
-14.8 -20.0 -5.0 0.0 
297.0 16.0 320.0 0.0 
0.0 6.0 -30.0 -36.3 


### Вектор b моего варианта: 

In [4]:
print("%.1f "*len(b) % tuple(b) )

943.0 -80.7 2602.8 1.1 


### Решение с помощью linalg.solve

In [5]:
x_np = np.linalg.solve(A, b)
print("%.4f "*len(x_np) % tuple(x_np) )

6.2355 -1.1964 2.4063 -2.2167 


# Метод Зейделя. 

Метод Зейделя - итерационный метод решения систем линейных алгебраических уравнений.
Метод Зейделя похож на метод Якоби, но главное отличие - новое значение $x^{(k+1)}$ используется сразу, как только его нашли.

Основная идея: 
1. Система $ Ax = b $ проеобразуется к виду $ x=Bx+c $. Данный вид необходим для создания итераций.
2. Матрица В представляется как $ B = B_1 + B_2 $,  
где $B_1$ - нижнетреугольная матрица, $B_2$ - верхнетреугольная матрица. На диагонали обоих этих матриц стоят нули.

Итерации алгоритма определены следующим образом. 

$$x^{(k+1)} = B_1 x^{(k+1)} + B_2 x^{(k)} +c $$

### Преобразование системы 𝐴𝑥=𝑏 к итерационному виду x=Bx+c


Используется следующее преобразование для получения матрицы В и вектора с: 

$ b_{ij} = 0$, если i=j

$b_{ij} = -\frac {a_{ij}}{a_{ii}} $, если $ i \neq j $

$c_i =\frac{b_i}{a_{ii}} $

In [6]:
B= []
c= []

for i in range(len(b)):
    c.append(b[i]/A[i][i])
    B.append([])
    for j in range(len(b)):
        if(i==j):
            B[i].append(0)
        else:
            B[i].append(-A[i][j]/A[i][i])
            
 

### Полученная матрица В:

In [7]:
for i in range(len(B)):
    print("%.3f "*len(B[i]) % tuple(B[i]) )
 

0.000 0.118 0.042 0.750 
-0.742 0.000 -0.250 -0.000 
-0.928 -0.050 0.000 0.000 
-0.000 0.165 -0.826 0.000 


### Полученный вектор с:

In [8]:
print("%.3f "*len(c) % tuple(c) )

7.938 4.035 8.134 -0.030 


In [9]:
B_norm = np.linalg.norm(B, np.inf)
print("Норма полученной матрицы В:")
print(B_norm)

Норма полученной матрицы В:
0.9924999999999999


Достаточное условие сходимости: $$ ||B||_{\infty} < 1 $$ 


Данное условие выполняется для полученной матрицы В. 

### Реализация метода Зейделя 

B - полученная ранее матрица для системы вида x=Bx+c


$$ B = B_1 + B_2 $$ 
$B_1$ - нижнетреугольная матрица, $B_2$ - верхнетреугольная матрица.

$$x^{(k+1)} = B_1 x^{(k+1)} + B_2 x^{(k)} +c $$

In [10]:
def zeid(B,c,iterations,x_start  ):
    #возвращает массив - вектор решений 
    y=x_start.copy()
   
    for k in range(iterations):
        x_k = y.copy()      
        for i in range(len(c)):
            u=0
            for j in range(i):
                u=u+B[i][j]*y[j]  #использование новых значений x^{k+1}          
            for j in range(i+1, len(c)):
                u=u+B[i][j]*x_k[j]          
            y[i] = u+c[i]
        print("iteration:" ,k+1)
        print("%.3f "*len(y) % tuple(y) )
        
    return y    
    

Выполним 10 итераций по методу Зейделя. 
Начальное приближение: $ x_0 = (1; 1; 1; 1)$

In [11]:
x = zeid(B, c, 10, [1,1,1,1])

iteration: 1
8.848 -2.784 0.061 -0.541 
iteration: 2
7.206 -1.331 1.512 -1.500 
iteration: 3
6.720 -1.332 1.964 -1.873 
iteration: 4
6.458 -1.251 2.202 -2.057 
iteration: 5
6.340 -1.223 2.310 -2.142 
iteration: 6
6.284 -1.209 2.362 -2.182 
iteration: 7
6.258 -1.202 2.385 -2.200 
iteration: 8
6.246 -1.199 2.396 -2.209 
iteration: 9
6.240 -1.198 2.402 -2.213 
iteration: 10
6.238 -1.197 2.404 -2.215 


Абсолютная погрешность для найденного вектора решений:
$$ \Delta x=||x_{from linalg}-x_{zeidel}||_{\infty} $$

In [12]:
abs_error  =   np.linalg.norm(x_np-x, np.inf)
print("Абсолютная погрешность: ", abs_error)

Абсолютная погрешность:  0.00233976886560594


Относительная погрешность для вектора решений: 
 $$ \delta( x) = \frac{||x_{from linalg}-x_{zeidel}||_{\infty} }{ || x_{from linalg}||_{\infty}  } $$

In [13]:
relative_error  =   np.linalg.norm( x_np - x, np.inf)/ np.linalg.norm(x_np, np.inf)
print("Относительная погрешность: ", relative_error)

Относительная погрешность:  0.00037523640238665545


Ранее полученное решение: $x \approx (6.2355; -1.1964; 2.4063; -2.2167 ) $ 

Как видно, решение написанной программы сходится к истинному решению.

Выполним снова 10 итераций по методу Зейделя. 
Начальное приближение: $ x_0 = (10; 5; 6; 100)$

In [14]:
x = zeid(B, c, 10, [10,5,6,100])

iteration: 1
83.779 -59.671 -66.640 45.181 
iteration: 2
31.987 -3.055 -21.402 17.152 
iteration: 3
19.541 -5.124 -9.746 7.178 
iteration: 4
12.307 -2.666 -3.155 2.137 
iteration: 5
9.093 -1.928 -0.210 -0.176 
iteration: 6
7.570 -1.533 1.185 -1.263 
iteration: 7
6.860 -1.355 1.835 -1.771 
iteration: 8
6.527 -1.270 2.139 -2.008 
iteration: 9
6.372 -1.231 2.281 -2.119 
iteration: 10
6.299 -1.213 2.348 -2.171 


In [15]:
abs_error  =   np.linalg.norm(x_np-x, np.inf)
print("Абсолютная погрешность: ", abs_error)

Абсолютная погрешность:  0.06386664858270574


In [16]:
relative_error  =   np.linalg.norm( x_np - x, np.inf)/ np.linalg.norm(x_np, np.inf)
print("Относительная погрешность: ", relative_error)

Относительная погрешность:  0.010242503778449479


Решение итерационного метода снова сходится к истинному решению. Это объясняется тем, что выполняется достаточное условие сходимости метода: $$ ||B|| < 1 $$ 
Значит, метод будет сходиться при любом начальном приближении.

Во втором случае значения погрешностей выше, так как вектор начального приближения был выбран дальше от истинного решения, при этом число итераций было одинаковым. 

### Поиск решения с заданной точностью 

 $$ \varepsilon = 10^{-6} $$
 
 Пусть $||B_1||+||B_2||<1. $ Тогда используется следующий критерий окончания: 

 
$$ ||x^{(k+1)} - x^{(k)}|| <  \frac{1-||B||}{||B_2||} \varepsilon$$

Сначала получим матрицы В1 и В2 для проверки выполнения.

In [17]:
B1=[]
B2=[]
for i in range(len(B)):
    B1.append([])
    B2.append([])
    for j in range(len(B[0])):
        if(i==j):
            B1[i].append(0)
            B2[i].append(0)
        elif(i>j):
            B1[i].append(B[i][j])
            B2[i].append(0)
        else:
            B2[i].append(B[i][j])
            B1[i].append(0)            

In [18]:
print("Матрица  В1: ")
for i in range(len(B1)):
    print("%.3f "*len(B1[i]) % tuple(B1[i]) )
print("Матрица  В2: ")
for i in range(len(B2)):
    print("%.3f "*len(B2[i]) % tuple(B2[i]) )
  

Матрица  В1: 
0.000 0.000 0.000 0.000 
-0.742 0.000 0.000 0.000 
-0.928 -0.050 0.000 0.000 
-0.000 0.165 -0.826 0.000 
Матрица  В2: 
0.000 0.118 0.042 0.750 
0.000 0.000 -0.250 -0.000 
0.000 0.000 0.000 0.000 
0.000 0.000 0.000 0.000 


In [19]:
norm_B1 = np.linalg.norm( B1, np.inf)
norm_B2 = np.linalg.norm( B2, np.inf)
print("Сумма норм В1 и В2: ", norm_B1+norm_B2)

Сумма норм В1 и В2:  1.9016681971227425


Будем использовать упрощенный критерий окончания итерационного процесса Зейделя:
   $$ ||x^{(k+1)} - x^{(k)}||  < \varepsilon_2 $$

In [27]:
def zeid_with_precision(B,B1,B2,c,x_start, epsilon = 1e-6  ):
     
    x_start = np.array(x_start)
    k=1
    y=np.zeros(4)
    
    #первая итерация 
    for i in range(len(c)):
        u=0
        for j in range(len(c)):
            u=u+B[i][j]*x_start[j]            
                   
        y[i] = u+c[i]        
        
    compare_norm = np.linalg.norm( y-x_start, np.inf)
    q = (1 - np.linalg.norm( B, np.inf) ) / np.linalg.norm(B2, np.inf)
    epsilon2 = epsilon*q
    while(compare_norm>epsilon2):
   
        x_k = y.copy()      
        for i in range(len(c)):
            u=0
            for j in range(i):
                u=u+B[i][j]*y[j]            
            for j in range(i+1, len(c)):
                u=u+B[i][j]*x_k[j]          
            y[i] = u+c[i]
        compare_norm = np.linalg.norm( y-x_k, np.inf)
        k=k+1
        
    print("Number of iterations: ", k )
    return y    
    

In [28]:
x_new = zeid_with_precision(B,B1,B2, c, [10,1,1,1])

Number of iterations:  27


### Полученный вектор решений: 

In [29]:
print("%.9f "*len(x_new) % tuple(x_new) )

6.235452778 -1.196396237 2.406290202 -2.216724063 


In [30]:
abs_error  =   np.linalg.norm(x_np-x_new, np.inf)
print("Абсолютная погрешность: ", abs_error)

Абсолютная погрешность:  5.389856561066608e-09


Для поиска решения с точностью  $ \varepsilon = 10^{-6} $ потребовалось 27 итераций метода Зейделя. 